src/ui/src/js/services/garden_service.js

```diff
@@ -89,22 +116,184 @@ export default function gardenService($rootScope, $http) {
+      !headersZeroLength &&
+      headerEntry['headers'].every(isEmptyObject)
+    );
+    const headersAllBlank = (
+      !headersAllEmpty &&
+      headerEntry['headers'].every(
+          (entry) =>
+            ('key' in entry && isBlank(entry['key'])) ||
+            ('value' in entry && isBlank(entry['value'])))
+    );
+
+    return (
+      !headersExist ||
+      headersZeroLength ||
+      headersAllEmpty ||
+      headersAllBlank
+    );
+  };
+
+  // eslint-disable-next-line no-unused-vars
+  const cleanEmptyStompHeaders = () => 'TODO';
+
+  const isEmptyStompConnection = (entryPointValues) => {
+    /* If every field is missing, then obviously the stomp connection can be
+     * considered empty.
+     *
+     * It gets a little more complicated in other cases because a lot of garbage
+     * data is being passed around (an issue for another day).
+     *
+     * So we do a lot of checking of the corner cases in
+     * isEmptyStompConnection and isEmptyStompHeaders so that the results of
+     * this function is truly representative of whether we would consider the
+     * connection to be "empty".
+     *
+     * (The point of all this is that if the connection meets our common-sense
+     * definition of empty, then the resulting connection parameter object
+     * won't even have a 'stomp' entry at all, which is far preferable to
+     * polluting the database with the cruft that gets picked up in the UI.)
+     */
+    const simpleFieldMissing = getSimpleFieldPredicate(entryPointValues);
+
+    const stompSimpleFields = [
+      'host', 'password', 'port', 'send_destination', 'subscribe_destination',
+      'username',
+    ];
+    const stompSslFields = ['ca_cert', 'client_cert', 'client_key'];
+
+    const allSimpleFieldsMissing = stompSimpleFields.every(simpleFieldMissing);
+    const headersMissing = isEmptyStompHeaders(entryPointValues);
+    const sslIsMissing = typeof entryPointValues['ssl'] === 'undefined' ||
+        !!Object.entries(entryPointValues['ssl']);
+    let nestedFieldsMissing = true;
+
+    if (!sslIsMissing) {
+      nestedFieldsMissing = stompSslFields.every(
+          (entry) =>
+            typeof entryPointValues['ssl'][entry] === 'undefined' ||
+            entryPointValues['ssl'][entry] === null ||
+            entryPointValues['ssl'][entry] === '',
+      );
+    }
+
+    const allStompFieldsEmpty = headersMissing && allSimpleFieldsMissing &&
+        nestedFieldsMissing;
+
+    return allStompFieldsEmpty;
+  };
+
+  const isEmptyHttpConnection = (entryPointValues) => {
+    // Simply decide if every field in the http entry is blank.
+    const simpleFieldMissing = getSimpleFieldPredicate(entryPointValues);
+    const httpSimpleFields = [
+      'ca_cert', 'client_cert', 'host', 'port', 'url_prefix',
+    ];
+    const allHttpFieldsEmpty = httpSimpleFields.every(simpleFieldMissing);
+
+    return allHttpFieldsEmpty;
+  };
+
+  GardenService.formToServerModel = function(data, model) {
```

Comment:
I haven't been able to find a scenario where all of this additional checking does what I think you're intended for it to do, at least in terms of it's potential usefulness to a user.  A few things I've tried:

1) Create a new garden, select "STOMP" as the connection type, change nothing else, click save.  It saved just fine, despite there being no values set for any of the STOMP parameters.  This is consistent with the existing behavior, but seemed like something these additions were aiming to prevent.
2) Clear all of the fields of the HTTP config, select HTTP as the connection type.  Click Export.  Nothing happens.  There's an error logged to the console, but nothing communicated to the user.
3) Same scenario, click "Save configuration".  All of the boxes get the green checkmark (despite being empty) making it appear as if things got validated and saved.  In reality, nothing happened and there's a warning message at the top saying as much, but I would have to scroll up to see that now.

It's not clear to me that the additional validation here is actually providing value, so I'd suggest we just drop it.  I'm open to hearing arguments though if you really thing there's something here.

Reply:
Fixed. It was intentionally throwing an exception, but the exception handling code was removed with the validation code. Now the exception throwing code is removed as well and it should act consistently across the scenarios you describe.